### PR TITLE
NimbleEdge and ONNXRuntime Blog June 17

### PR DESCRIPTION
**Pull request by MaanavD**
DRAFT PR until images and thumbnail are provided

### CODE REVIEW

**Comment by MaanavD** [comment]
https://maanavd.github.io/onnxruntime/blogs

^staged here

File: src/routes/blogs/nimbleedge-x-onnxruntime/+page.svx [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+Traditionally, recommendation systems utilized only historical customer behavior (e.g. purchases, views, likes), updated weekly, daily, or every few hours. However, pioneering apps such as [LinkedIn](https://www.linkedin.com/blog/engineering/member-customer-experience/near-real-time-features-for-near-real-time-personalization), [Doordash](https://doordash.engineering/2021/03/04/building-a-declarative-real-time-feature-engineering-framework/), and [Pinterest](https://medium.com/pinterest-engineering/how-pinterest-leverages-realtime-user-actions-in-recommendation-to-boost-homefeed-engagement-volume-165ae2e8cde8), have lately incorporated near-real-time and real-time features in recommendations. This has enabled highly fresh recommendations based on user interactions from just a few minutes, or even seconds ago, enabling massive uptick in key business metrics.
```
Comment (skottmckay) [inline src/routes/blogs/nimbleedge-x-onnxruntime/+page.svx:36]: 'enabling massive uptick in key business metrics' -> 'enabling a massive uptick in key business metrics'?

File: src/routes/blogs/nimbleedge-x-onnxruntime/+page.svx
Comment (skottmckay) [inline src/routes/blogs/nimbleedge-x-onnxruntime/+page.svx:1]: There's a lot of usage of 'prominent'/'prominently' but not sure if we should be editing overuse of specific words with synonyms.

File: src/routes/blogs/nimbleedge-x-onnxruntime/+page.svx [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+## IMAGE HERE
+
+However, building real-time ML system on cloud involves several major challenges:
+
+- **High Compute Costs:** Prohibitively large cloud costs for storing and processing in-session user interaction data in real-time
+- **Time to market and complexity:** Setting up real-time cloud pipelines for ML is complex, and requires significant time and developer bandwidth
+- **Privacy Risk and Compliance:** Users' in-session clickstream data must be sent to cloud servers for processing, raising privacy concerns
+- **Scaling Challenges:** Handling traffic surges requires either massive reserve cloud capacity, or rapid scaling which is difficult to achieve on cloud
+- **High Latency:** High latency due to time required to make cloud API calls from/ to user mobile devices
```
Comment (skottmckay) [inline src/routes/blogs/nimbleedge-x-onnxruntime/+page.svx:48]:
```suggestion
- **High Latency:** High latency due to time required to make cloud API calls from/to user mobile devices
```

File: src/routes/blogs/nimbleedge-x-onnxruntime/+page.svx [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+- **Privacy:** In-session user interaction data is processed on device, enhancing privacy
+- **Latency:** Most modern smartphones are highly capable of rapid data processing and inference computations
+- **Scaling:** Effortless scaling, as traffic increase translates to more mobile devices available for computation
+
+However, developing and maintaining on-device real-time ML systems is also highly complex, and requires building systems for
+
+1. on-device real-time event capture and storage
+2. on-device feature processing (like rolling window aggregates)
+3. feature syncing from cloud
+4. ML inference execution suitably balancing latency and resource utilization across diverse mobile devices.
```
Comment (skottmckay) [inline src/routes/blogs/nimbleedge-x-onnxruntime/+page.svx:64]: Blog doesn't show bullet points or numbering for this list which seems inconsistent. Should we convert from a numbered list to dashes?

File: src/routes/blogs/nimbleedge-x-onnxruntime/+page.svx [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+All of these are highly complex and time taking endeavors, due to which performant on-device real-time ML is highly challenging.
```
Comment (skottmckay) [inline src/routes/blogs/nimbleedge-x-onnxruntime/+page.svx:66]:
```suggestion
All of these are highly complex and time consuming endeavors, due to which performant on-device real-time ML is highly challenging.
```

File: src/routes/blogs/nimbleedge-x-onnxruntime/+page.svx [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+#### **Ready-to-use on-device ML infrastructure**
+
+- **On-device data warehouse:** Store in-session user events (e.g. product views, cart additions in e-commerce) in a managed on-device database, with querying latency `< 1ms`
+- **Data processing:** Run Python-like preprocessing scripts on-device to compute features from real-time user inputs
+- **Feature store synchronization:** Maintain a frequently updated on-device copy of cloud features required for real-time ML (e.g. list of open restaurants for real-time ranking of restaurants on a food delivery app's homepage)
+
+#### **Effortless on-device ML model management**
+
+- **Over the air updates,** to complete decouple ML model updates and front-end app updates, enabling aid rapid ML experimentation
```
Comment (skottmckay) [inline src/routes/blogs/nimbleedge-x-onnxruntime/+page.svx:80]:
```suggestion
- **Over the air updates,** to completely decouple ML model updates and front-end app updates, enabling aid rapid ML experimentation
```

File: src/routes/blogs/nimbleedge-x-onnxruntime/+page.svx [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+For inference execution, NimbleEdge utilizes a number of runtimes, prominently including ONNX Runtime. [ONNX Runtime](https://onnxruntime.ai/) is an open-source, high-performance engine to accelerate machine learning models on both cloud and edge environments. One of its key features, ONNX Runtime Mobile, is specifically optimized to deliver lighter and faster model inferences. It offers configurations for **quantization, hardware acceleration (e.g. using CPU/GPU/NPU), operator fusion, parallel processing,** and more. NimbleEdge leverages ONNX Runtime Mobile in a variety of impactful ways:
+
+- Applying static and dynamic quantization techniques to significantly reduce the memory footprint of models
+- Customizing the runtime build to match the specific operators required by the model(s)
```
Comment (skottmckay) [inline src/routes/blogs/nimbleedge-x-onnxruntime/+page.svx:94]:
```suggestion
- Customizing the runtime build to match the specific operators and data types required by the model(s)
```

File: src/routes/blogs/nimbleedge-x-onnxruntime/+page.svx (hunk 3 above)
Comment (MaanavD) [reply src/routes/blogs/nimbleedge-x-onnxruntime/+page.svx:64]: Fixed, thanks :)

File: src/routes/blogs/nimbleedge-x-onnxruntime/+page.svx (hunk 1 above)
Comment (EmmaNingMS) [reply src/routes/blogs/nimbleedge-x-onnxruntime/+page.svx]: Updated.

File: src/routes/blogs/nimbleedge-x-onnxruntime/+page.svx
Comment (EmmaNingMS) [reply src/routes/blogs/nimbleedge-x-onnxruntime/+page.svx:1]: We have completed the marketing review. Will leave them as it.

**Review by EmmaNingMS — APPROVED** [review]

**Review by EmmaNingMS — APPROVED** [review]

**Pull request merged**